### PR TITLE
Logic collision

### LINKED ISSUE #28: Move collision logic out of Entities and into Logic
It is not good that each entity keeps track on how it interacts with all other entities.
It is much better if Logic can keep track of it, especially since a collision is a relation between two entities, not one.

A solution should be flexible enough to allow defining general rules (Bullet hits Enemy) which applies to all interactions between various sub-classes that does not have overriding rules.

### PR DESCRIPTION
**Pull request by zephyyrr**
This fixes #28. Finally.

Rules can be divided up into packs, like in the case of "Modern".
Logic adds whatever rules it wants into its CollisionRules implementation and that lets it handle applying the correct rules to the correct entities.

I envision that in the future, it is no longer Logic doing the actual entities.find(...) call to find collisions, but a CollisionFinder kind of system to easily change between this simple system and a possible Quad-tree implementation.

### CODE REVIEW

File: src/Game.js [hunk 1]
```diff
@@ -40,6 +40,7 @@ export class Game {
 
   tick(delta) {
     this.update(delta);
+    this.logic.checkCollisions(this, this.entities);
```
Comment (Victorystick) [inline src/Game.js]: Can you move this into the `update` method just before where we remove "dead" entities instead? We'll want to be able to remove entities after a collision has occurred.

File: src/Logic.js [hunk 2]
```diff
@@ -1,11 +1,15 @@
...
+import {CollisionRules} from './collision/CollisionRules.js';
+import {Modern} from './collision/modern.js';
 
 export class Logic {
   constructor() {
     this.running = true;
     this.level = 0;
     this.timeSinceSpawn = 0;
+    this.collision = new CollisionRules();
+    Modern.initialize(this.collision);
```
Comment (Victorystick) [inline src/Logic.js]: Inject this in the constructor instead? That way we don't have to alter the Logic class to change the collision behaviour.

File: src/collision/CollisionRules.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    this.collisions.push((game, e1, e2) => {
+      if (e1 instanceof type1 && e2 instanceof type2) {
+        return collision_fn(game, e1, e2);
+      }
+      return true;
+    });
+  }
+
+  onCollide(game, entity1, entity2) {
+    for (var fn of this.collisions) {
```
Comment (Victorystick) [inline src/collision/CollisionRules.js]: `var` -> `const`

File: src/collision/CollisionRules.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  constructor() {
+    this.collisions = new Array();
+  }
+
+  add(type1, type2, collision_fn) {
+    this.collisions.push((game, e1, e2) => {
+      if (e1 instanceof type1 && e2 instanceof type2) {
+        return collision_fn(game, e1, e2);
+      }
+      return true;
```
Comment (Victorystick) [inline src/collision/CollisionRules.js]: What does it mean for the function to return `true` or `false`?

File: src/collision/CollisionRules_test.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
+import assert from 'assert';
+import {createArgSaver} from '../testing/mocks.js';
+import {CollisionRules} from './CollisionRules.js';
+import {Vec2, Size} from '../core.js';
+import {Enemy, Bullet, PlasmaBullet} from '../entities/all.js';
+
+describe('CollisionRules', () => {
+  describe('onCollide', () => {
+    const enemy = new Enemy(new Vec2(0,0), new Size(0,0), null, 0, 0);
+
```
Comment (Victorystick) [inline src/collision/CollisionRules_test.js]: We also want a test case where the collision function is not called. It's currently possible to pass these tests with an implementation that doesn't even include any `instanceof` checks

**Comment by Victorystick** [comment]
It looks good besides my comments. I'll take another look once they're addressed.

**Comment by zephyyrr** [comment]
The thing with returning true or false is if the collision system should look for more handlers or not.
So true to look for more and false for stop.

I guess it could be made more explicit with a call to the expected "super" handler, but that puts dependencies on rules which we should strive to not have.

Another solution would be a more general object return with a "continue" or "super" property set.
But i figured until we know more of what we need from this system, this was good enough.

**Comment by Victorystick** [comment]
LGTM!

**Comment by zephyyrr** [comment]
If it LGTY, then I shall merge it and finally be done with the issue #28 that has stopped so much progress!

**Pull request merged**